src/v/pandaproxy/kafka_client_cache.cc

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "pandaproxy/kafka_client_cache.h"
+
+#include "pandaproxy/logger.h"
+
+#include <seastar/core/loop.hh>
+
+namespace pandaproxy {
+
+kafka_client_cache::kafka_client_cache(
+  YAML::Node const& cfg,
+  config::binding<size_t> max_size,
+  config::binding<std::chrono::milliseconds> keep_alive)
+  : _config{cfg}
+  , _cache_max_size{max_size}
+  , _keep_alive{keep_alive} {}
+
+client_ptr kafka_client_cache::make_client(
+  credential_t user, config::rest_authn_method authn_method) {
+    kafka::client::configuration cfg{
+      to_yaml(_config, config::redact_secrets::no)};
+
+    // Set the principal when the request is using HTTP Basic AuthN
+    if (authn_method == config::rest_authn_method::http_basic) {
+        // Need to specify type or else bad any_cast runtime error
+        cfg.sasl_mechanism.set_value(ss::sstring{"SCRAM-SHA-256"});
+        cfg.scram_username.set_value(std::move(user.name));
+        cfg.scram_password.set_value(std::move(user.pass));
+    }
+
+    return ss::make_lw_shared<kafka::client::client>(
+      to_yaml(cfg, config::redact_secrets::no));
+}
+
+ss::future<client_ptr> kafka_client_cache::fetch_or_insert(
+  credential_t user, config::rest_authn_method authn_method) {
+    auto& inner_list = _cache.get<underlying_list>();
+    auto& inner_hash = _cache.get<underlying_hash>();
+    size_t k{_secret_hash(user.name + user.pass)};
```

Comment:
I can't think of a scenario where we'd want to have multiple clients for a given user with a different password. I guess the old one will timeout at some point.

Reply:
@dotnwat brought a good use case to my attention in https://github.com/redpanda-data/redpanda/pull/6452#discussion_r978221159

This hash accounts for when a user changes their password after their first round of requests. I also like the idea of hashing by a keyed integer instead of a string.